src/hooks/resources/device.ts

```diff
@@ -579,3 +595,105 @@ async function checkSupervisorReleaseUpgrades(
 		}
 	}
 }
+
+async function setSupervisorReleaseResource(
+	api: sbvrUtils.PinejsClient,
+	deviceIds: number[],
+	supervisorVersion: string,
+) {
+	if (deviceIds.length === 0) {
+		return;
+	}
+	const devices = (await api.get({
+		resource: 'device',
+		options: {
+			$expand: {
+				is_of__device_type: {
+					$select: 'id',
+				},
+			},
+			// if the device already has a supervisor_version, just bail.
+			$filter: {
+				id: { $in: deviceIds },
+				supervisor_version: { $eq: null },
+			},
+			$select: ['id', 'supervisor_version'],
+		},
+	})) as AnyObject;
+
+	if (devices.length === 0) {
+		return;
+	}
+
+	const deviceTypes = _.uniqBy(
+		_.reduce(
+			devices,
+			(result: AnyObject[], val, _key) => {
+				// TODO: hardcoded index
+				result.push(val.is_of__device_type[0].id);
+				return result;
+			},
+			[],
+		),
+		'id',
+	);
+
+	if (deviceTypes.length === 0) {
+		return;
+	}
+
+	await Bluebird.map(deviceTypes, async (deviceType) => {
+		const affectedDeviceIds = _.map(devices, (d) => {
+			// TODO: hardcoded index
+			if (d.is_of__device_type[0].id === deviceType) {
+				return d.id;
+			}
+		});
```

Comment:
Take a look at [_.groupBy](https://lodash.com/docs/4.17.15#groupBy) - you should be able to handle this and the `uniqBy` above in one go with
```ts
const devicesByDeviceType = _.groupBy(device, (d) => d.is_of__device_type[0].id)
```
and get a `{ [deviceTypeId: number]: device[] }` mapping back

Reply:
this be much cleaner, thanks